codegen/realtime.py

```diff
@@ -0,0 +1,291 @@
+		if m[1]:
+			lua_type = type_to_lua(m[1])
+			name = m[2]
+			repeated = m[0] == "repeated "
+			if repeated:
+				lua_type == "table"
+			properties.append({ "type": lua_type, "name": name, "repeated": repeated})
+	return properties
+
+
+def message_to_lua(message_id, api):
+	message = get_proto_message(message_id, api)
+	if not message:
+		print("Unable to find message %s" % message_id)
+		return
+
+	props = parse_proto_message(message)
+	function_args = [ "socket" ]
+	for prop in props:
+		function_args.append(prop["name"])
+	function_args.append("callback")
+
+	function_args_string = ", ".join(function_args)
+
+	message_id = camel_to_snake(message_id)
+	function_name = "send_" + message_id
+
+	lua = "\n"
+	lua = lua + "--- " + function_name + "\n"
+	for function_arg in function_args:
+		lua = lua + "-- @param %s\n" % (function_arg)
+	lua = lua + "function M.%s(%s)\n" % (function_name, function_args_string)
+	lua = lua + "	assert(socket)\n"
+	for prop in props:
+		lua = lua + "	assert(_G.type(%s) == '%s')\n" % (prop["name"], prop["type"])
+	lua = lua + "	local message = {\n"
+	lua = lua + "		%s = {\n" % message_id
+	for prop in props:
+		lua = lua + "			%s = %s,\n" % (prop["name"], prop["name"])
+	lua = lua + "		}\n"
+	lua = lua + "	}\n"
+	lua = lua + "	return socket_send(socket, message, callback)\n"
+	lua = lua + "end\n"
+	return lua
+
+
+def event_to_lua(event_id, api):
+	event = get_proto_message(event_id, api)
+	if not event:
+		print("Unable to find event %s" % event_id)
+		return
+	
+	event_id = camel_to_snake(event_id)
+	function_name = "on_" + event_id
+
+	lua = "\n"
+	lua = lua + "--- " + function_name + "\n"
+	lua = lua + "-- @param socket Nakama Client Socket.\n"
+	lua = lua + "-- @param fn The callback function.\n"
+	lua = lua + "function M.%s(socket, fn)\n" % (function_name)
+	lua = lua + "	assert(socket, \"You must provide a socket\")\n"
+	lua = lua + "	assert(fn, \"You must provide a function\")\n"
+	lua = lua + "	socket.events.%s = fn\n" % (event_id)
+	lua = lua + "end\n"
+	return { "name": function_name, "lua": lua }
+
+
+if len(sys.argv) == 0:
+	print("You must provide both an input file")
+	sys.exit(1)
+
+proto_path = sys.argv[1]
+out_path = None
+
+if len(sys.argv) > 2:
+	out_path = sys.argv[2]
+
+
+
+api = read_as_string(proto_path)
+
+
+CHANNEL_MESSAGES = [ "ChannelJoin", "ChannelLeave", "ChannelMessageSend", "ChannelMessageRemove" ]
+MATCH_MESSAGES = [ "MatchData", "MatchDataSend", "MatchJoin", "MatchLeave" ]
+MATCHMAKER_MESSAGES = [ "MatchmakerAdd", "MatchmakerRemove" ]
+PARTY_MESSAGES = [ "PartyCreate", "PartyJoin", "PartyLeave", "PartyPromote", "PartyAccept", "PartyRemove", "PartyClose", "PartyJoinRequestList", "PartyMatchmakerAdd", "PartyMatchmakerRemove", "PartyDataSend" ]
+STATUS_MESSAGES = [ "StatusFollow", "StatusUnfollow", "StatusUpdate" ]
+ALL_MESSAGES = CHANNEL_MESSAGES + MATCH_MESSAGES + MATCHMAKER_MESSAGES + PARTY_MESSAGES + STATUS_MESSAGES
+
+
+CHANNEL_EVENTS = [ "ChannelPresenceEvent" ]
```

Comment:
List of incoming events/messages that will have listener functions generated (in the format `on_message_name`). Are these all? I think maybe also Match should be added?